Editorial review: Add information on file system access locking modes (#35334)
mdn/content: files/en-us/web/api/filesystemfilehandle/createsyncaccesshandle/index.md

```js-nolint
createSyncAccessHandle()
createSyncAccessHandle(options)
```

### Parameters

- `options` {{optional_inline}}

  - : An object with the following properties:

    - `mode` {{optional_inline}} {{non-standard_inline}}
      - : A string specifying the locking mode for the access handle. The default value is `"readwrite"`.
        Possible values are:
        - `"read-only"`

>           - : Multiple `FileSystemSyncAccessHandle`s can be opened simultaneously on a file (for example when using the same app in multiple tabs), provided they are all opened in `"read-only"` mode. Once opened, read-like methods can be called on the handles — {{domxref("FileSystemSyncAccessHandle.read", "read()")}}, {{domxref("FileSystemSyncAccessHandle.getSize", "getSize()")}}, and {{domxref("FileSystemSyncAccessHandle.close", "close()")}}.

Author: Fixed in my next commit.

        - `"readwrite"`
          - : Only one `FileSystemSyncAccessHandle` can be opened simultaneously. Attempting to open subsequent handles before the first handle is closed results in a `NoModificationAllowedError` exception being thrown. Once opened, any available method can be called on the handle.
        - `"readwrite-unsafe"`

>           - : Multiple `FileSystemSyncAccessHandle`s can be opened simultaneously on a file, provided they are all opened in `"readwrite-unsafe"` mode. Once opened, any available method can be called on the handles.

Author: fixed

};
```

### Complete example with `mode` option

> Our [createSyncAccessHandle() mode test](https://createsyncaccesshandle-mode-test.glitch.me/) example provides an {{htmlelement("input")}} field to enter text into, and two buttons — one to write entered text to the end of a file in the origin private file system, and one to empty the file when it becomes too full.

Reviewer: When I test this site, I don't see anything after `File contents:` when I write to the file.  Looks like it might be an encoding issue because I can see some replacement characters when I paste it here: `File contents:     
`

Author: Darnit, this is really annoying. I've used this exact technique successfully in the past. I am seeing the same result as you, including the replacement characters.

https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/TextDecoder#fatal implies that the text decoder is receiving malformed data, but I'm not sure how to fix this. Have you got any ideas?

Reviewer: It looks like you need to pass the at option to read: `accessHandle.read(dataView, { at: 0 });`

Author: Ah, Brilliant, thanks @nathanmemmott, that has done the trick. Updated, in both the glitch and the code on the page.

> Our [createSyncAccessHandle() mode test](https://createsyncaccesshandle-mode-test.glitch.me/) example provides an {{htmlelement("input")}} field to enter text into, and two buttons — one to write entered text to the end of a file in the origin private file system, and one to empty the file when it becomes too full.

Author: updated

> Try exploring the demo above, with the browser developer console open so you can see what is happening. If you try opening the demo in multiple browsers (get some friends to help!) you will find that multiple handles can be opened at once to write to the file at the same time. This is because `mode: "readwrite-unsafe"` is set on the `createSyncAccessHandle()` calls.

Author: updated

Below we'll explore the code.

#### HTML

The two {{htmlelement("button")}} elements and text {{htmlelement("input")}} field look like this:

```html
<ol>
  <li>
    <label for="filetext">Enter text to write to the file:</label>
    <input type="text" id="filetext" name="filetext" />
  </li>
  <li>
    Write your text to the file: <button class="write">Write text</button>
  </li>
  <li>
    Empty the file if it gets too full:
    <button class="empty">Empty file</button>
  </li>
</ol>
```

#### Main JavaScript

The main thread JavaScript inside the HTML file is shown below. We grab references to the write text button, empty file button, and text input field, then we create a new web worker using the {{domxref("Worker.Worker", "Worker()")}} constructor. We then define two functions and set them as event handlers on the buttons:

- `writeToOPFS()` is run when the write text button is clicked. This function posts the entered value of the text field to the worker using the {{domxref("Worker.postMessage()")}} method, then empties the text field, ready for the next addition.

> - `emptyOPFS()` is run when the empty file button is clicked. This posts a very specific string to the worker that we will use to identify when the file is to be emptied rather then written to.

Author: updated

```js
const writeBtn = document.querySelector(".write");
const emptyBtn = document.querySelector(".empty");
const fileText = document.querySelector("#filetext");

const opfsWorker = new Worker("worker.js");

function writeToOPFS() {
  opfsWorker.postMessage(fileText.value);
  console.log("Text posted to worker");
  fileText.value = "";
}

function emptyOPFS() {

>   opfsWorker.postMessage("!!empty!!");

Author: Heh, I was being lazy; you are totally right ;-)

I'll update this in the demo and the text.

}

writeBtn.addEventListener("click", writeToOPFS);
emptyBtn.addEventListener("click", emptyOPFS);
```

#### Worker JavaScript

The worker JavaScript is shown below.

> First, we run a function called `initOPFS()` that gets a reference to the OPFS root using {{domxref("StorageManager.getDirectory()")}}, creates a file and returns its handle using {{domxref("FileSystemDirectoryHandle.getFileHandle()")}}, and then returns an accessHandle using `createSyncAccessHandle()` with `mode: "readwrite-unsafe"` set to allow multiple handles to access the same file simultaneously.

Reviewer 2: ```suggestion
First, we run a function called `initOPFS()` that gets a reference to the OPFS root using {{domxref("StorageManager.getDirectory()")}}, creates a file and returns its handle using {{domxref("FileSystemDirectoryHandle.getFileHandle()")}}, and then returns a `FileSystemSyncAccessHandle` using `createSyncAccessHandle()` with `mode: "readwrite-unsafe"` set to allow multiple handles to access the same file simultaneously.
```

Author: updated

```js
let accessHandle;

async function initOPFS() {
  const opfsRoot = await navigator.storage.getDirectory();
  const fileHandle = await opfsRoot.getFileHandle("file.txt", { create: true });
  accessHandle = await fileHandle.createSyncAccessHandle({
    mode: "readwrite-unsafe",
  });
}

initOPFS();
```

> Inside the worker's [message event](/en-US/docs/Web/API/Worker/message_event) handler function, we first get the size of the file using {{domxref("FileSystemSyncAccessHandle.getSize", "getSize()")}}. We then check to see whether the data sent in the message is the specific string sent by the `emptyOPFS()` method — `"!!empty!!"`. If so, we empty the file using {{domxref("FileSystemSyncAccessHandle.truncate", "truncate()")}} with a vlaue of `0`, and update the file size contained in the `size` variable.

Reviewer 2: If you follow my recommendation above, the text mentioning the microsyntax needs to be adapted.

Author: yup, updated.

If the message data is something else, we:

- Create a new {{domxref("TextEncoder")}} and {{domxref("TextDecoder")}} to handle encoding and decoding the text content later on.
- Encode the message data and write the result to the end of the file using {{domxref("FileSystemSyncAccessHandle.write", "write()")}}, then update the file size contained in the `size` variable.
- Create a {{domxref("DataView")}} to contain the file contents, and read the content into it using {{domxref("FileSystemSyncAccessHandle.read", "read()")}}.
- Decode the `DataView` contents and log it to the console.

```js
onmessage = function (e) {
  console.log("Worker: Message received from main script");

  // Initialize this variable for the size of the file

>   let size;

Reviewer 2: ```suggestion
```

You can do this below.

Author: good call; more concise, but it also makes the code quicker and easier to read.

  // Get the current size of the file

>   size = accessHandle.getSize();

Reviewer 2: ```suggestion
 let size = accessHandle.getSize();
```

Author: as above, updated
